python/tskit/tables.py

```diff
@@ -2470,3 +2470,172 @@ def drop_index(self):
+
+        1. Nodes: if in the list of nodes, and in the order provided.
+        2. Individuals and Populations: if referred to by a retained node,
+            and in the order first seen when traversing the list of retained nodes.
+        3. Edges: if both parent and child are retained nodes.
+        4. Mutations: if the mutations' node is a retained node.
+        5. Sites: if any mutations remain at the site after removing mutations.
+        6. Migrations: if the migration's node is a retained node.
+
+        If ``nodes`` is the entire list of nodes in the tables, then the resulting
+        tables will be identical to the original tables, but with nodes
+        (and individuals and populations) reordered.
+
+        **Note:** This is quite different from :meth:`.simplify`: the resulting
+        tables contain only the nodes given, not ancestral ones as well, and
+        does not simplify the relationships in any way.
+
+        :param list[int] nodes: A list of node IDs to retain in the tables.
+        """
+        if nodes.shape[0] == 0:
+            raise ValueError("Nodes cannot be empty.")
+        if np.max(nodes) >= self.nodes.num_rows:
+            raise ValueError("One of the nodes is not in the TableCollection.")
+        tables = self.copy()
+        n = tables.nodes
+        # figuring out which individuals to keep
+        old_inds = n.individual[nodes]
+        indiv_to_keep, i = np.unique(old_inds, return_index=True)
+        # maintaining order in which they appeard
+        indiv_to_keep = indiv_to_keep[np.argsort(i)]
+        # removing -1
+        indiv_to_keep = indiv_to_keep[indiv_to_keep != tskit.NULL]
+        # subsetting individuals table
+        if indiv_to_keep.shape[0] == 0:
+            self.individuals.clear()
+        else:
+            i = tables.individuals
+            self.individuals.set_columns(
+                flags=i.flags[indiv_to_keep],
+                **util.subset_ragged_col(
+                    i.location, i.location_offset, indiv_to_keep, "location"
+                ),
+                **util.subset_ragged_col(i.metadata, i.metadata_offset, indiv_to_keep),
+            )
+        # figuring out which pops to keep
+        old_pops = n.population[nodes]
+        pop_to_keep, j = np.unique(old_pops, return_index=True)
+        pop_to_keep = pop_to_keep[np.argsort(j)]
+        pop_to_keep = pop_to_keep[pop_to_keep != tskit.NULL]
+        # subsetting populations table
+        if pop_to_keep.shape[0] == 0:
+            self.populations.clear()
+        else:
+            p = tables.populations
+            self.populations.set_columns(
+                **util.subset_ragged_col(p.metadata, p.metadata_offset, pop_to_keep)
+            )
+        # mapping of ind/pop id in full tables to subset tables
+        ind_map = {ind: i for i, ind in enumerate(indiv_to_keep)}
+        pop_map = {pop: j for j, pop in enumerate(pop_to_keep)}
```

Comment:
perhaps set `ind_map[tskit.NULL] = tskit.NULL`? (also, you should use NULL, not -1)

Reply:
and, with that in, I wonder if `np.fromiter(map(ind_map.get, old_inds))` would be more efficient? not quite as clear, though.

Reply:
Or, wouldn't it be simpler to use the same strategy as `node_map` below, with an array instead of a dict?

Reply:
no, because the individual and population tables could be empty, in which case numpy version would fail

Reply:
also, according to [this](https://stackoverflow.com/questions/35215161/most-efficient-way-to-map-function-over-numpy-array/35216364) not clear if you gain much from using `np.fromiter(map)` compared to a comprehension.